### PR TITLE
ENH: 5x speed increase for staypoint generation

### PR DESCRIPTION
**Pull request by bifbof**
I tinkered around a bit and looked at the code of the staypoint generation as that takes always a lot of time. 
Found out that most of the time is spent accessing the x and y coordinates (shapely is slow) so I have thrown the coordinates into numpy arrays. That way the access takes nanoseconds instead of microseconds.
I also moved some calculation outside of the for-loop to speedup the time comparisons but those improvements are rather minor.
Now the limiting factor is the distance function. For this reason I replaced the casting to radians with `np.deg2rad` as this is some microseconds faster.

This results in a speed up of around a factor 5. (Tested on my laptop with users 51 and 52)
- Before the changes it took around 52 seconds
- After the changes it takes around 11 seconds

If we want to increase the speed of the function further we probably have to use either cython or numba.
Another way would be to do the distance calculations in "blocks" instead of single entries that way we can use the speedup from numpy more.

### CODE REVIEW

File: trackintel/preprocessing/positionfixes.py [hunk 1]
```diff
@@ -415,78 +415,72 @@ def _generate_staypoints_sliding_user(
...
-    pfs = df.to_dict("records")
-    idx = df.index.to_list()
+    df = df.sort_index(kind="stable").sort_values(by=["tracked_at"], kind="stable")
+    # transform times to pandas Timedelta to simplify comparisons
+    gap_threshold = pd.Timedelta(gap_threshold, unit="minutes")
+    # precalulate variables to not do that in for loop
+    time_threshold = time_threshold * 60
+    last_curr = len(df.index) - 1
+    # to numpy as access time of numpy numpy is faster than pandas array
+    gap_times = pd.eval("((df.tracked_at - df.tracked_at.shift(1)) > gap_threshold)").to_numpy()
```
Comment (hong2223) [inline trackintel/preprocessing/positionfixes.py:426]: Nice solution!

**Review by hong2223 — APPROVED** [review]
Thanks for the nice solution! This is a really important update.

I tested with the whole geolife dataset and this PR will create a major speedup!

File: trackintel/preprocessing/positionfixes.py [hunk 2]
```diff
@@ -415,78 +415,72 @@ def _generate_staypoints_sliding_user(
...
 
-    df = df.sort_index(kind="mergesort").sort_values(by=["tracked_at"], kind="mergesort")
-    # pfs id should be in index, create separate idx for storing the matching
-    pfs = df.to_dict("records")
-    idx = df.index.to_list()
+    df = df.sort_index(kind="stable").sort_values(by=["tracked_at"], kind="stable")
+    # transform times to pandas Timedelta to simplify comparisons
+    gap_threshold = pd.Timedelta(gap_threshold, unit="minutes")
+    # precalulate variables to not do that in for loop
+    time_threshold = time_threshold * 60
```
[the remark this answers is not in the document]
Comment (bifbof) [reply trackintel/preprocessing/positionfixes.py]: Yes I think it also fine like this, I just wanted to move the calculation outside of the function.

File: trackintel/preprocessing/positionfixes.py [hunk 3]
```diff
@@ -416,11 +416,12 @@ def _generate_staypoints_sliding_user(
         raise AttributeError("distance_metric unknown. We only support ['haversine']. " f"You passed {distance_metric}")
 
     df = df.sort_index(kind="stable").sort_values(by=["tracked_at"], kind="stable")
-    # transform times to pandas Timedelta to simplify comparisons
-    gap_threshold = pd.Timedelta(gap_threshold, unit="minutes")
```
Comment (bifbof) [inline trackintel/preprocessing/positionfixes.py]: Does the comparison still work if we delete that?

**Pull request merged**